### PR TITLE
MRG: update build matrix for Python 3.6

### PR DESCRIPTION
**Pull request by matthew-brett**
* Add Python 3.6 tests;
* Switch away from un-maintained travis-wheels wheelhouse to pypi + nipy server
  (for Python 3.6 matplotlib wheels).

Reveals test failures for Python 3.6 and with --pre flag.

### CODE REVIEW

**Comment by matthew-brett** [comment]
Error for ``--pre`` flag may be because of change to pypi to look for packages.   For example, the ``--pre`` option picks up `matplotlib==2.0.0rc2` for the configuration in this PR (see [here](https://travis-ci.org/matthew-brett/scikit-image/jobs/191828434#L497)), but previous builds, using the old `travis-wheels` wheelhouse for packages, were picking up `matplotlib==1.5.1` (see [here](https://travis-ci.org/scikit-image/scikit-image/jobs/191742083#L485)).

**Comment by matthew-brett** [comment]
Yup - the problem is with matplotlib 2.0.0rc2.   In the failing example : https://github.com/scikit-image/scikit-image/blob/master/doc/examples/edges/plot_circular_elliptical_hough_transform.py#L135

the `edges` array is of dtype `bool`, although the script seems to think it's of type `uint8`.  Matplotlib 2.0.0rc2 rejects an `imshow` on the `bool` array, insisting on an uint8, uint16 or float.

I can make the error go away by casting `edges` to `uint8` - but I am not sure that's the right fix.  Can someone take a look?

**Comment by sciunto** [comment]
@matthew-brett your suggestion sounds good to me. 
Thanks for looking at this.

**Comment by JDWarner** [comment]
Hrm, `imshow` on a boolean array is actually a very common use case, e.g., for viewing or overlaying segmented objects.  I'll look into why this limitation was made; perhaps it can be lifted before 2.0 final.

**Comment by matthew-brett** [comment]
Josh - the error is from using a boolean RGB array - so the red, green, blue channels all have to be encoded with booleans.   I haven't thought too hard about it, but superficially I can see that might be a sign that you had the wrong datatype.

**Comment by matthew-brett** [comment]
Although I suppose you could sensibly interpret the boolean RGB array elements as encoding one of black, white, red, green, blue, red+green, red+blue, green+blue).

**Comment by sciunto** [comment]
Yes, we had a wrong datatype (I pledge guilty, I wrote that as one of my first contribution to scikit-image). :)

**Comment by sciunto** [comment]
looks like we have other errors... I get to go.

**Comment by matthew-brett** [comment]
Yup - am investigating now, will feed back shortly.

**Comment by matthew-brett** [comment]
OK - not having much fun here.

The error on macOS seems to be intermittent, caused by a bad URL to the tex sources: https://travis-ci.org/scikit-image/scikit-image/jobs/191989789#L6630 ; try searching https://s3.amazonaws.com/archive.travis-ci.org/jobs/191989789/log.txt for "ctan" for errors.

The Python 3.6 error may be to do with an incompatibility with networkx on Python 3.6 - see https://travis-ci.org/scikit-image/scikit-image/jobs/191989788#L3516 and #2450. 

I don't understand the error for the optional dependencies build:

```
======================================================================
ERROR: skimage.io.tests.test_imageio.test_imageio_truncated_jpg
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/venv/lib/python3.4/site-packages/PIL/ImageFile.py", line 97, in __init__
    self._open()
  File "/home/travis/venv/lib/python3.4/site-packages/PIL/JpegImagePlugin.py", line 310, in _open
    i = i8(s)
  File "/home/travis/venv/lib/python3.4/site-packages/PIL/_binary.py", line 24, in i8
    return c if c.__class__ is int else c[0]
IndexError: index out of range
```

https://travis-ci.org/scikit-image/scikit-image/jobs/191989785#L8270

I couldn't replicate this on on OSX with Python 3.4 and imageio installed.   This run uses imageio (2.1.1).  A [comparable passing run](https://travis-ci.org/scikit-image/scikit-image/jobs/191742090) on master uses imageio (1.3) - so it might be a version incompatibility.

Does anyone have any ideas?

**Comment by matthew-brett** [comment]
OK, tracked down and fixed the imageio error - see commit 3ddff7f . I am out of my depth for the graph error on Python 3.6.   Doing dumb debugging, the error for `assert new_labels.max() == 0` is because `np.all(new_labels) == 1` (on Python 3.6).

The difference might be in the output of networkx.  Saving the generated graph `rag` from `test_ncut_stable_subgraph` as a pickle allows me to load the Python 3.6 graph in Python 3.5 and vice versa.  Call the Python 3.6 generated graph `rag36` and the other `rag35`.  Python 3.5 returns all zero labels from `graph.cut_normalized(labels, rag, in_place=False)` regardless of which of `rag35`, `rag36` it is operating on.   Python 3.6 returns all zero labels from `rag35` but all 1 labels from `rag36`.    I gave up at that point as I really don't know networkx or graph algorithms.

**Comment by matthew-brett** [comment]
@vighneshbirodkar - I see this is your graph test - would you mind taking a look?

**Comment by matthew-brett** [comment]
Networkx failing it's own test suite on Python 3.6: https://github.com/networkx/networkx/issues/2350

File: tools/travis_before_install.sh [hunk 1]
```diff
@@ -5,11 +5,28 @@ set -ex
 export COVERALLS_REPO_TOKEN=7LdFN9232ZbSY3oaXHbQIzLazrSf6w2pQ
 export PIP_DEFAULT_TIMEOUT=60
 
+# This URL is for any extra wheels that are not available on pypi.  As of 14
+# Jan 2017, the major packages such as numpy and matplotlib are up for all
+# platforms.  The URL comes points ot a Rackspace CDN belonging to the
```
Comment (soupault) [inline tools/travis_before_install.sh]: Something is wrong here: `The URL comes points ot`

**Comment by soupault** [comment]
Are we going to merge this making all the tests for Python 3.6 to fail? (see https://github.com/scikit-image/scikit-image/issues/2424)

File: tools/travis_before_install.sh (hunk 1 above)
Comment (matthew-brett) [inline tools/travis_before_install.sh]: Thanks - fixed.

**Comment by matthew-brett** [comment]
For networkx - they're working on Python 3.6 tests over at https://github.com/networkx/networkx/pull/2319 . Most of the changes are adapting the tests to the new dictionary ordering in Python 3.6.  For the skimage code, I guess we need to establish whether the Python 3.6 version of the graph is also correct, and adapt the test accordingly.

I tried running the skimage test against the networkx PR above, but unfortunately the networkx API has changed, so I got lots of errors of form:

```
TypeError: add_edge() takes 3 positional arguments but 4 were given
```

I made a PR to fix up compatibility with the new networkx API here : https://github.com/scikit-image/scikit-image/pull/2455

Using those changes, and the Python 3.6 networkx PR, I still see the same `new_labels.max()` test error over at #2450, so I suspect that the graph returned in Python 3.6 is correct, and it's our test assumptions that are incorrect, but that's only a guess, I don't have enough expertise to go further.

**Comment by matthew-brett** [comment]
Why don't I comment out the Python 3.6 matrix entry for now, and uncomment when #2450 is addressed?

**Comment by soupault** [comment]
>Why don't I comment out the Python 3.6 matrix entry for now, and uncomment when #2450 is addressed?

@matthew-brett Yes, I think that would be a good solution. I'm fine whether we have a single red cross in the panel or not, but the other core devs might go hesitant (we don't want to scare them away from reviewing :wink: ) .
In any case, what you are doing for fixing 3.6 compatibility is priceless!

At the moment, we are fixing the remaining issues for 0.13. I'm going to take care of #2450 right after the release.

**Comment by matthew-brett** [comment]
@soupault - but I guess you'd want to do Python 3.6 wheels for the new release?

**Comment by matthew-brett** [comment]
Meanwhile numpy 1.12 came along and caused some new errors - added new commit to fix ...

**Comment by matthew-brett** [comment]
OK - tests all passing now (except for 3.6, where tests commented out).  Ready to merge from my  point of view.

**Comment by soupault** [comment]
@matthew-brett Wait... Why did you reference #2450? :thinking: Actually, I was speaking about https://github.com/scikit-image/scikit-image/issues/2424 all along (it causes CI to fail on Py 3.6).

> but I guess you'd want to do Python 3.6 wheels for the new release?

Taking into consideration our release cycle, I don't think we should wait till 0.14, and build them asap (in the ideal case - right after #2424 is fixed). We can do this with the next minor release (i.e. 0.13.1).

Not so many packages claim to support Py 3.6 at the moment, but I already have it in my distro, and the incompatibility causes some issues.

**Comment by soupault** [comment]
@matthew-brett This PR is now blocking: https://github.com/scikit-image/scikit-image/pull/2451 , https://travis-ci.org/scikit-image/scikit-image/jobs/192504528 .

**Comment by matthew-brett** [comment]
@soupault - sorry - I don't understand what you mean by "This PR is now blocking ...".   Looking at the error log you pointed to, this PR should fix that error - no?

For Python 3.6 wheels - I think these are urgent, as Python 3.6 is now the default download for OSX on Python.org.   All the major packages now have Python 3.6 wheels - cython, numpy, scipy, matplotlib, pandas, scikit-learn ...

**Comment by matthew-brett** [comment]
I think we should merge this one - any objections?

**Comment by soupault** [comment]
@matthew-brett 
>sorry - I don't understand what you mean by "This PR is now blocking ...". Looking at the error log you pointed to, this PR should fix that error - no?

Yes, sorry, I mean this PR (2451) has to be merged before the mentioned one. Due to some caching magic, it appears that `sphinx` in our CI uses Python 3.6 now.

I'm :+1: to merge this PR, but we need a second reviewer.

---

>For Python 3.6 wheels - I think these are urgent

I fully agree.
I'd volunteer to transfer the testing environment to `pytest`, but that is a matter of 1-2 weeks (actually, 2-3 including the review)..

**Comment by matthew-brett** [comment]
But nosetests seems to run fine for Python 3.6 - what is the reason to wait for the port to py.test ?

**Comment by stefanv** [comment]
Thank you very much for all the work done here; I was starting to follow the same spiral this morning, and then discovered that you had already patched all the issues along the way.

**Comment by sciunto** [comment]
@matthew-brett The deprecation warning says that some stuffs will be broken with python 3.6, but it seems it's not. Nevertheless, we should not sit on that :)

**Comment by matthew-brett** [comment]
@sciunto - absolutely - sounds like the move to py.test is a good idea, but it shouldn't be a blocker for Python 3.6 support.

**Comment by soupault** [comment]
Ah, indeed, they are kept as warnings only.
:+1: to enable Py3.6 then (with the fix to https://github.com/scikit-image/scikit-image/issues/2450, I assume).

```
[egor@vbox scikit-image]$ python skimage/transform/tests/test_integral.py
/usr/lib/python3.6/importlib/_bootstrap.py:205: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/home/egor/Workspace/scikit-image/skimage/feature/blob.py:178: DeprecationWarning: invalid escape sequence \s
  """
/home/egor/Workspace/scikit-image/skimage/feature/blob.py:286: DeprecationWarning: invalid escape sequence \s
  """
...EE
======================================================================
ERROR: skimage.transform.tests.test_integral.test_vectorized_integrate
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/egor/Workspace/scikit-image/skimage/transform/tests/test_integral.py", line 46, in test_vectorized_integrate
    assert_equal(expected, integrate(s, r0, c0, r1, c1))  # test deprecated
  File "/home/egor/Workspace/scikit-image/skimage/transform/integral.py", line 86, in integrate
    warn("The syntax 'integrate(ii, r0, c0, r1, c1)' is "
  File "/home/egor/Workspace/scikit-image/skimage/_shared/_warnings.py", line 16, in warn
    warnings.warn(message, stacklevel=stacklevel)
UserWarning: The syntax 'integrate(ii, r0, c0, r1, c1)' is deprecated, and will be phased out in release 0.14. The new syntax is 'integrate(ii, (r0, c0), (r1, c1))'.

======================================================================
ERROR: test suite for <module 'skimage.transform.tests.test_integral' from '/home/egor/Workspace/scikit-image/skimage/transform/tests/test_integral.py'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/nose/suite.py", line 228, in run
    self.tearDown()
  File "/usr/lib/python3.6/site-packages/nose/suite.py", line 351, in tearDown
    self.teardownContext(ancestor)
  File "/usr/lib/python3.6/site-packages/nose/suite.py", line 367, in teardownContext
    try_run(context, names)
  File "/usr/lib/python3.6/site-packages/nose/util.py", line 453, in try_run
    inspect.getargspec(func)
  File "/usr/lib/python3.6/inspect.py", line 1041, in getargspec
    DeprecationWarning, stacklevel=2)
DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()

----------------------------------------------------------------------
Ran 4 tests in 0.526s

FAILED (errors=2)
```

**Pull request merged**